lib/utils/compose.js

```diff
@@ -0,0 +1,1086 @@
+							return pushAndUpdateServiceImages(docker, token, images, function(
+								serviceImage,
+							) {
+								logger.logDebug(
+									`Saving image ${serviceImage.is_stored_at__image_location}`,
+								);
+								if (skipLogUpload) {
+									delete serviceImage.build_log;
+								}
+								return releaseMod.updateImage(
+									client,
+									serviceImage.id,
+									serviceImage,
+								);
+							});
+						})
+						.finally(function() {
+							logger.logDebug('Untagging images...');
+							return Promise.map(images, ({ localImage }) =>
+								localImage.remove(),
+							);
+						});
+				})
+				.then(() => {
+					release.status = 'success';
+				})
+				.tapCatch(() => {
+					release.status = 'failed';
+				})
+				.finally(function() {
+					runloop = runSpinner(tty, spinner, `${prefix}Saving release...`);
+					release.end_timestamp = new Date();
+					if (release.id == null) {
+						return;
+					}
+					return releaseMod
+						.updateRelease(client, release.id, release)
+						.finally(runloop.end);
+				})
+				.return(release);
+		});
+}
+
+// utilities
+
+const renderProgressBar = function(percentage, stepCount) {
+	const _ = require('lodash');
+	percentage = _.clamp(percentage, 0, 100);
+	const barCount = Math.floor((stepCount * percentage) / 100);
+	const spaceCount = stepCount - barCount;
+	const bar = `[${_.repeat('=', barCount)}>${_.repeat(' ', spaceCount)}]`;
+	return `${bar} ${_.padStart(percentage, 3)}%`;
+};
+
+var pushProgressRenderer = function(tty, prefix) {
+	const fn = function(e) {
+		const { error, percentage } = e;
+		if (error != null) {
+			throw new Error(error);
+		}
+		const bar = renderProgressBar(percentage, 40);
+		return tty.replaceLine(`${prefix}${bar}\r`);
+	};
+	fn.end = () => {
+		tty.clearLine();
+	};
+	return fn;
+};
+
+var createLogStream = function(input) {
+	const split = require('split');
+	const stripAnsi = require('strip-ansi-stream');
+	return input.pipe(stripAnsi()).pipe(split());
+};
+
+var dropEmptyLinesStream = function() {
+	const through = require('through2');
+	return through(function(data, _enc, cb) {
+		const str = data.toString('utf-8');
+		if (str.trim()) {
+			this.push(str);
+		}
+		return cb();
+	});
+};
+
+var buildLogCapture = function(objectMode, buffer) {
+	const through = require('through2');
+
+	return through({ objectMode }, function(data, _enc, cb) {
+		// data from pull stream
+		if (data.error) {
+			buffer.push(`${data.error}`);
+		} else if (data.progress && data.status) {
+			buffer.push(`${data.progress}% ${data.status}`);
+		} else if (data.status) {
+			buffer.push(`${data.status}`);
+
+			// data from build stream
+		} else {
+			buffer.push(data);
+		}
+
+		return cb(null, data);
+	});
+};
+
+var buildProgressAdapter = function(inline) {
+	const through = require('through2');
+
+	const stepRegex = /^\s*Step\s+(\d+)\/(\d+)\s*: (.+)$/;
+
+	let step = null;
+	let numSteps = null;
+	let progress;
```

Comment:
Changed this from being assigned via a constant destructure to being individually assigned, previously was
```
[step, numSteps, progress] = [null, null, undefined]